cmd/openshift-install/graph.go

```diff
@@ -0,0 +1,95 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"reflect"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/awalterschulze/gographviz"
+	"github.com/spf13/cobra"
+
+	"github.com/openshift/installer/pkg/asset"
+)
+
+var (
+	graphOpts struct {
+		outputFile string
+	}
+)
+
+func newGraphCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "graph",
+		Short: "Outputs the internal dependency graph for installer",
+		Long:  "",
+		RunE:  runGraphCmd,
+	}
+	cmd.PersistentFlags().StringVar(&graphOpts.outputFile, "output-file", "", "file where the graph is written, if empty prints the graph to Stdout.")
+	return cmd
+}
+
+func runGraphCmd(cmd *cobra.Command, args []string) error {
+	g := gographviz.NewGraph()
+	g.SetName("G")
+	g.SetDir(true)
+	g.SetStrict(true)
+
+	tNodeAttr := map[string]string{
+		string(gographviz.Shape): "box",
+		string(gographviz.Style): "filled",
+	}
+	for _, t := range targets {
+		name := fmt.Sprintf(`"%s"`, t.name)
+		g.AddNode("G", name, tNodeAttr)
+		for _, dep := range t.assets {
+			addEdge(g, name, dep)
+		}
+	}
+
+	out := os.Stdout
+	if graphOpts.outputFile != "" {
+		f, err := os.Create(graphOpts.outputFile)
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+		out = f
+	}
+
+	if _, err := io.WriteString(out, g.String()); err != nil {
+		return err
+	}
+	return nil
+}
+
+func addEdge(g *gographviz.Graph, parent string, asset asset.Asset) {
+	elem := reflect.TypeOf(asset).Elem()
+	name := strings.TrimPrefix(fmt.Sprintf(`"%s/%s"`, elem.PkgPath(), elem.Name()), `"github.com/openshift/installer/`)
```

Comment:
The trim leaves this without the required leading quote.  And I'd [still][1] prefer context via [`URL`][2].  Something like:

```dot
InstallConfig [URL="https://godoc.org/github.com/openshift/installer/pkg/asset/installconfig#InstallConfig"]
```

For private classes, maybe just link to the package?

To make it easier to tune this, maybe add a new commit updating our dependency SVG and dropping our current hand-coded dot source?


[1]: https://github.com/openshift/installer/pull/432#discussion_r223767145
[2]: https://graphviz.gitlab.io/_pages/doc/info/attrs.html#d:URL

Reply:
dropped the name to just name of elem for now. we can iterate over the url stuff later.
Also adding the commit for replacing existing graph in docs.